bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/io/RefreshingOAuth2CredentialsInterceptor.java

```diff
@@ -187,155 +199,135 @@ public RefreshingOAuth2CredentialsInterceptor(ExecutorService scheduler,
     synchronized (lock) {
-      return !isRefreshing && getCacheState(this.headerCache.get()) != CacheState.Good;
-    }
-  }
-
-  private static StatusRuntimeException asUnauthenticatedException(Exception e) {
-    return Status.UNAUTHENTICATED.withCause(e).asRuntimeException();
-  }
-
-  @VisibleForTesting
-  static CacheState getCacheState(HeaderCacheElement headerCache) {
-    return (headerCache == null) ? CacheState.Expired : headerCache.getCacheState();
-  }
-
-  private HeaderCacheElement getCachedHeader() throws IOException {
-    HeaderCacheElement headerCache = this.headerCache.get();
-    if (headerCache != null && headerCache.exception != null) {
-      throw headerCache.exception;
-    }
-    return headerCache;
-  }
-
-  /**
-   * Perform a credentials refresh.
-   */
-  @VisibleForTesting
-  boolean doRefresh() {
-    if (!canRefresh()) {
-      return false;
-    }
-    synchronized (lock) {
-      if (!canRefresh()) {
-        return false;
+      if (isRefreshing) {
+        return futureToken;
       }
       isRefreshing = true;
-    }
-    try {
-      HeaderCacheElement cacheElement = refreshCredentialsWithRetry();
-      synchronized (lock) {
-        updateToken(cacheElement);
-      }
-    } finally {
-      synchronized (lock) {
+
+      try {
+        this.futureToken = executor.submit(new Callable<HeaderCacheElement>() {
+          @Override
+          public HeaderCacheElement call() throws Exception {
+            HeaderCacheElement newToken = refreshCredentialsWithRetry();
+            return updateToken(newToken);
+          }
+        });
+      } catch (RuntimeException e) {
         isRefreshing = false;
-        lock.notifyAll();
+        throw e;
       }
+
+      return this.futureToken;
     }
-    return true;
   }
 
   private HeaderCacheElement updateToken(HeaderCacheElement newToken) {
-    // Update the token only if the new token is good or the old token is bad
-    CacheState newState = newToken.getCacheState();
-    boolean newTokenOk = newState == CacheState.Good || newState == CacheState.Stale;
-    CacheState oldCacheState = getCacheState(headerCache.get());
-    boolean oldTokenOk = oldCacheState == CacheState.Good || oldCacheState == CacheState.Stale;
-
-    if (newTokenOk || !oldTokenOk) {
-      headerCache.set(newToken);
-      return newToken;
-    } else {
-      LOG.warn("Failed to refresh the access token. Falling back to existing token. "
-          + "New token state: {}", newState);
-      return headerCache.get();
+    synchronized (lock) {
+      // Update the token only if the new token is good or the old token is bad
+      CacheState newState = newToken.getCacheState();
+      boolean newTokenOk = newState == CacheState.Good || newState == CacheState.Stale;
+      CacheState oldCacheState = headerCache.getCacheState();
+      boolean oldTokenOk = oldCacheState == CacheState.Good || oldCacheState == CacheState.Stale;
+
+      if (newTokenOk || !oldTokenOk) {
```

Comment:
Unless I made a mistake, we would only use the old stale token if the new token is an error, which allows us to continue making requests.